fendermint/app/src/app.rs

```diff
@@ -490,6 +529,47 @@ where
         Ok(response)
     }
 
+    /// 1. Prepare txns
+    /// 2. Prepare parent finality proposal
+    async fn prepare_proposal(
+        &self,
+        request: request::PrepareProposal,
+    ) -> AbciResult<response::PrepareProposal> {
+        let max_tx_bytes: usize = request.max_tx_bytes.try_into().unwrap();
+        let mut size: usize = 0;
+        let mut txs = Vec::new();
+        for tx in request.txs {
+            if size.saturating_add(tx.len()) > max_tx_bytes {
+                break;
+            }
+            size += tx.len();
+            txs.push(tx);
+        }
+
+        if let Some(finality_proposal) = self
+            .parent_finality
+            .next_finality_proposal()
+            .map_err(|e| anyhow!("cannot get next finality to propose due to: {e}"))?
+        {
+            let bytes =
+                serde_json::to_vec(&ChainMessage::IPC(IPCMessage::TopDown(finality_proposal)))?;
```

Comment:
Once moved into the `ChainMessageInterpreter` you won't have to convert to binary (not JSON) because that is take care of by the `BytesMessageInterpreter`. 

In any case, it was wrong to put this _after_ the stuff above that summed up all the bytes, because this could have brought the total above the allowed limit. This will not be a problem if you rebase onto https://github.com/consensus-shipyard/fendermint/pull/185 because there the check is done after the interpreters have finished. 

For this reason (the possibility of this check removing a message we wanted to include in the proposal) it is crucial that for example `next_finality_proposal()` actually does things in a read-only fashion, which is why I indicated in the `ProposalInterpreter` that it does not return a modified state (it's a `Reader` monad, not a `State` monad :nerd_face: ), because our proposal might not be voted on, might time out, or may be removed. We have to be prepared to re-propose it. Only the transaction execution should cause changes, e.g. when we see that our proposal went through, then we can move on to the next candidate.

If we used STM, we might actually want to maintain two regions there: a "transient" region modified during `deliver_tx`, and a "permanent" region where it's merged during `commit`.